rangefeedcache,settingswatcher,systemcfgwatcher: lose gossip deps

ajwerner opened the pull request:
This is a pile of commits which supersedes https://github.com/cockroachdb/cockroach/pull/69269 and pretty much puts in place all of the pieces to move on https://github.com/cockroachdb/cockroach/issues/70560.

ajwerner committed:
rangefeedcache,spanconfig/kvsubscriber: extract rangefeed cache logic

This logic is generally handy when one wants to build caching atop rangefeeds
and wants to make claims about the data corresponding to a consistent snapshot.

Release note: None

ajwerner committed:
server/settingswatcher: adopt rangefeedcache

This commit doesn't do anything in terms of functional changes, but it lays
the groundwork for a subsequent change to buffer and write out a settings
snapshot.

Release note: None

ajwerner committed:
server/settingswatcher: add redaction and stringer support for RawValue

We log it.

Release note: None

ajwerner committed:
server/settingswatcher: adopt rangefeed for system tenant, remove gossip

This commit removes the code which connected the settings to their backing
table via the gossipped system config. Instead it unconditionally enables the
rangefeed-backed settingswatcher which was developed to support tenants.

Note that it is rather tested code that has been used in multi-tenant sql
pods for about a year now and all the existing tests still pass.

Release note: None

ajwerner committed:
kvclient/rangefeed/rangefeedbuffer: add some helpers for dealing with kvs

Often we want to extract KVs from buffered events and we want to merge them
to form a snapshot.

Release note: None

ajwerner committed:
kvclient/rangefeed/rangefeedcache: add a Cache for KVs

This will be handy when dealing with the system config and generally as a test
for this package.

Release note: None

ajwerner committed:
server/systemconfigwatcher: use rangefeeds to watch the system config

This also adopts the rangefeed backed cache for the optimizer and for
the reporter.

Fixes #70558
Fixes #74665

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74612)
<!-- Reviewable:end -->

ajwerner committed:
kvclient/rangefeed: add option to not set scan timestamp

Sometimes we want to know what was the timestamp in KV.

Release note: None

RaduBerinde reviewed (commented):
This is really great work!

My only concern is around the frontier bumping delay - I walked away from a previous discussion with the idea that the frontier lags behind the present by something on the order of seconds. That makes the decision of "upgrading" from a RangeFeed to a Watcher non-trivial, no? In particular, does that mean that `SET CLUSTER SETTING` now takes seconds?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-:-Mt5AOXACcvkj_vqnKYN:bnwwhgg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt5644yDX6528Qxn8MG:-Mt5644yDX6528Qxn8MH:bjr3xo1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/config.go#L111)):*
> ```Go
> }
> 
> // WithDiff makes an option to enable an initial scan which defaults to
> ```

This comment is incorrect (and had me scratching my head for a while).
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/kvs.go, line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt53UwdDFRneqwUjyd0:-Mt53UwdDFRneqwUjyd1:b-xorspf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2636978c6e53ec0be253ab86dd23b5b1e2dbe649/pkg/kv/kvclient/rangefeed/rangefeedbuffer/kvs.go#L20)):*
> ```Go
> 
> // RangeFeedValueEventToKV is a function to type assert an Event into a
> // *roachpb.RangeFeedValue and then conver it to a roachpb.KeyValue.
> ```

[nit] convert
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/kvs.go, line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt53Gug-2yGzlq-rpHf:-Mt53Gug-2yGzlq-rpHg:b350bba) ([raw file](https://github.com/cockroachdb/cockroach/blob/2636978c6e53ec0be253ab86dd23b5b1e2dbe649/pkg/kv/kvclient/rangefeed/rangefeedbuffer/kvs.go#L22)):*
> ```Go
> // *roachpb.RangeFeedValue and then conver it to a roachpb.KeyValue.
> func RangeFeedValueEventToKV(event Event) roachpb.KeyValue {
> 	return func(rfv *roachpb.RangeFeedValue) roachpb.KeyValue {
> ```

I don't understand what is going on here. How is this different than
```
rfv := event.(*roachpb.RangeFeedValue)
return roachpb.KeyValue{..}
```
?
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt529rt72kZ11C3RkPR:-Mt529rt72kZ11C3RkPS:blm4vye) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L40)):*
> ```Go
> 
> const (
> ```

[nit] blank line
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt50YQ29SypY1bCgol1:-Mt50YQ32P7o9Fhj99EL:bd82p2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L66)):*
> ```Go
> }
> 
> // Watcher is used to implement a consistent cache over spans of KV data
> ```

[supernit] it would be nice if this was the first thing in the file, easier to discover
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt52JHODizPzDmK1q2Y:-Mt52JHODizPzDmK1q2Z:b-iym8cb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L72)):*
> ```Go
> // loop. Rangefeeds used as is don't offer any ordering guarantees with
> // respect to updates made over non-overlapping keys, which is something we care
> // about. For that reason we inject logic to make use of a rangefeed buffer,
> ```

[nit] "we inject logic" can be misunderstood, I'd say "For that reason this facility implements logic to..."
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt58kp3-FVGlGxUMj7H:-Mt58kp3-FVGlGxUMj7I:b-m7bgcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L85)):*
> ```Go
> // The expectation is that the caller will use a mutex to update an underlying
> // data structure.
> type Watcher struct {
> ```

Shouldn't we include here the information that the events are received with a significant lag because of how the frontier is advanced?
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt50t0x9rR4YRtPQn5x:-Mt50t0x9rR4YRtPQn5y:blm4vye) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L106)):*
> ```Go
> // Knobs allows tests to inject behavior into the Watcher.
> type Knobs struct {
> ```

[nit] blank line
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt54oaQDvG1NhtNimBG:-Mt54oaQDvG1NhtNimBH:bfryq5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L124)):*
> ```Go
> 
> // NewWatcher instantiates a Watcher.
> func NewWatcher(
> ```

withPrevValue deserves some documentation here (point to `RangeFeedRequest.WithDiff`)
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt51w8G6mtHzL9zX2Hu:-Mt51w8G6mtHzL9zX2Hv:blm4vye) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L157)):*
> ```Go
> 		const aWhile = 5 * time.Minute // arbitrary but much longer than a retry
> 		for r := retry.StartWithCtx(ctx, base.DefaultRetryOptions()); r.Next(); {
> ```

[nit] blank line
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt516736zelT4o8W9aj:-Mt516736zelT4o8W9ak:b-gknndx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L180)):*
> ```Go
> // This is a blocking operation, returning only when the context is canceled,
> // or when a retryable error occurs. For the latter, it's expected that callers
> // will re-run the cache.
> ```

[nit] re-run the watcher?
___
*[pkg/server/settingswatcher/row_decoder.go, line 46 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt58Gm98r7G6Qrcvbox:-Mt58Gm98r7G6Qrcvboy:bxe7pj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/02996f4a34504e0f0c1cf50eda0d6d5bcee31879/pkg/server/settingswatcher/row_decoder.go#L46)):*
> ```Go
> }
> 
> func (r RawValue) SafeFormat(s interfaces.SafePrinter, verb rune) {
> ```

[nit] `// SafeFormat implements redact.SafeFormatter.`
___
*[pkg/server/settingswatcher/settings_watcher.go, line 127 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt5A3J4BPol_TycHcIt:-Mt5A3J4BPol_TycHcIu:bk9bakk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98fa7f2f4888e9eccd8f8e2f14e616e825ea94b/pkg/server/settingswatcher/settings_watcher.go#L127)):*
> ```Go
> 	}
> 
> 	c := rangefeedcache.NewWatcher(
> ```

Isn't there a significant functional change here in that the updates are now received with a multi-second delay? At the very least I can image this would make many tests much slower (if each time we set a cluster setting, we have to wait a few seconds)
___
*[pkg/server/settingswatcher/settings_watcher.go, line 130 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt59cOI-hLZMVhsXm99:-Mt59cOI-hLZMVhsXm9A:b-inn7aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98fa7f2f4888e9eccd8f8e2f14e616e825ea94b/pkg/server/settingswatcher/settings_watcher.go#L130)):*
> ```Go
> 		"settings-watcher",
> 		s.clock, s.f,
> 		0, // bufferSize
> ```

I don't understand what bufferSize=0 does? Wouldn't it cause an error on the first `onValue()` callback from the rangefeed?
___
*[pkg/server/settingswatcher/settings_watcher.go, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt58ejp6MztsjSqG2Al:-Mt58ejp6MztsjSqG2Am:bcnqhay) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98fa7f2f4888e9eccd8f8e2f14e616e825ea94b/pkg/server/settingswatcher/settings_watcher.go#L132)):*
> ```Go
> 		0, // bufferSize
> 		[]roachpb.Span{settingsTableSpan},
> 		false,
> ```

[nit] add comment
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt52sv00TK8fWwZhu4L:-Mt52sv00TK8fWwZhu4M:bojk971) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go#L142)):*
> ```Go
> 		ctx context.Context, v *roachpb.RangeFeedValue,
> 	) rangefeedbuffer.Event {
> 		return translateEvent(ctx, d, v)
> ```

[nit] should `translateEvent` be a method on `d` (in which case we would just use `d.translateEvent`)?
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go, line 289 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt52TtG4HDackb7lEmb:-Mt52TtG4HDackb7lEmc:b-15e9aa) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go#L289)):*
> ```Go
> }
> 
> func translateEvent(
> ```

[nit] could use a short comment
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go, line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt56niG30dkBt4GU7ys:-Mt56niG30dkBt4GU7yt:b-nbgzvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/af63f53d4db7d8ff673d143384e9ae0b5adc91d1/pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go#L83)):*
> ```Go
> 		updatedData = updateKVs
> 	case PartialUpdate:
> 		// Note that handleUpdate is not called synchronously, so we can use the
> ```

[nit] not called concurrently?
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt57ZWNDeehKjXudoZl:-Mt57ZWNDeehKjXudoZm:b-aqncmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/af63f53d4db7d8ff673d143384e9ae0b5adc91d1/pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go#L87)):*
> ```Go
> 		// missing anything.
> 		prev, _, _ := c.GetSnapshot() // okay if prev is nil
> 		if len(updateKVs) == 0 {
> ```

[nit] this special case could be handled in `MergeKVs`.
___
*[pkg/server/settingswatcher/row_decoder_test.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt584wD87WAvVZScWif:-Mt584wD87WAvVZScWig:b69wre3) ([raw file](https://github.com/cockroachdb/cockroach/blob/02996f4a34504e0f0c1cf50eda0d6d5bcee31879/pkg/server/settingswatcher/row_decoder_test.go#L1)):*
> ```Go
> package settingswatcher
> ```

[supernit] should remove this file from this commit


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> My only concern is around the frontier bumping delay - I walked away from a previous discussion with the idea that the frontier lags behind the present by something on the order of seconds. That makes the decision of "upgrading" from a RangeFeed to a Watcher non-trivial, no? In particular, does that mean that SET CLUSTER SETTING now takes seconds?

In the case of settings, it applies the settings change to the updater in the Translate function which is called before buffering the event. That means there is no delay for settings introduced by this change. It is a little cheeky to hijack the translate call like that. I’ll add commentary that this is a reasonable thing to do. In fact, if there’s no storage, it never even buffers the event, so it’s not really getting that much out of the watcher layer. The delay only affects the snapshot logic. 

Similarly, I don’t care about the delay very much for the two other use cases. 

One thing I’d like to do is deal with a hard restart of the watcher as will happen if the gc threshold is hit. 

I’m waiting at an urgent care, so I’ll address the feedback later. Thanks for taking a look. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-:-Mt5BxlO1p_PJeR52a4L:b-baws0k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)



<!-- Sent from Reviewable.io -->

ajwerner committed:
kvclient/rangefeed: fix WithDiff comment

Release note: None

ajwerner committed:
server/settingswatcher: handle rangefeedcache restarts

This is an improvement over what the library provided before. It previously
assumed that the rangefeed would never encounter a permanent error after the
initial scan. There are cases where such errors can occur. This commit defers
to the rangefeedcache.Start loop to handle them.

Release note: None

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74612)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/kv/kvclient/rangefeed/config.go, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt5644yDX6528Qxn8MG:-Mt6YGNG4a6grYUOMdn0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/config.go#L111)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This comment is incorrect (and had me scratching my head for a while).
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/kvs.go, line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt53UwdDFRneqwUjyd0:-Mt74v85DKaq7uK7meZG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2636978c6e53ec0be253ab86dd23b5b1e2dbe649/pkg/kv/kvclient/rangefeed/rangefeedbuffer/kvs.go#L20)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] convert
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedbuffer/kvs.go, line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt53Gug-2yGzlq-rpHf:-Mt6YOit8X2aTvN1wetp:b-evj5do) ([raw file](https://github.com/cockroachdb/cockroach/blob/2636978c6e53ec0be253ab86dd23b5b1e2dbe649/pkg/kv/kvclient/rangefeed/rangefeedbuffer/kvs.go#L22)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't understand what is going on here. How is this different than
```
rfv := event.(*roachpb.RangeFeedValue)
return roachpb.KeyValue{..}
```
?
</blockquote></details>

Done.

Bad refactor
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt529rt72kZ11C3RkPR:-Mt6ZiEXCuAxJpLnnKiR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L40)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] blank line
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt50YQ29SypY1bCgol1:-Mt6Ziyx3krf8oxGjPwU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L66)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[supernit] it would be nice if this was the first thing in the file, easier to discover
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt52JHODizPzDmK1q2Y:-Mt6ZjhZ9huL43nzoc8o:bcjdcq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L72)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] "we inject logic" can be misunderstood, I'd say "For that reason this facility implements logic to..."
</blockquote></details>

I reworded this comment.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt58kp3-FVGlGxUMj7H:-Mt6ZlzZAg98rqTQMYGG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L85)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Shouldn't we include here the information that the events are received with a significant lag because of how the frontier is advanced?
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt50t0x9rR4YRtPQn5x:-Mt6_PNE0AhukJbAPOVY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L106)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] blank line
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt54oaQDvG1NhtNimBG:-Mt6aZZT3r6nFvbTq3KX:bf7wyy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L124)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

withPrevValue deserves some documentation here (point to `RangeFeedRequest.WithDiff`)
</blockquote></details>

Added a bunch of commentary
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt51w8G6mtHzL9zX2Hu:-Mt6abUX052hxcjoM-kp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L157)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] blank line
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt516736zelT4o8W9aj:-Mt6aeI3CNbNeFZ9cCFp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L180)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] re-run the watcher?
</blockquote></details>

Done.
___
*[pkg/server/settingswatcher/row_decoder.go, line 46 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt58Gm98r7G6Qrcvbox:-Mt6afBHEpDgHtWCq6Pn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02996f4a34504e0f0c1cf50eda0d6d5bcee31879/pkg/server/settingswatcher/row_decoder.go#L46)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `// SafeFormat implements redact.SafeFormatter.`
</blockquote></details>

Done.
___
*[pkg/server/settingswatcher/settings_watcher.go, line 127 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt5A3J4BPol_TycHcIt:-Mt6cIk35LCVWY6srr6o:bl1qetl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98fa7f2f4888e9eccd8f8e2f14e616e825ea94b/pkg/server/settingswatcher/settings_watcher.go#L127)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Isn't there a significant functional change here in that the updates are now received with a multi-second delay? At the very least I can image this would make many tests much slower (if each time we set a cluster setting, we have to wait a few seconds)
</blockquote></details>

Nope. Hopefully the new commentary makes that clearer.
___
*[pkg/server/settingswatcher/settings_watcher.go, line 130 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt59cOI-hLZMVhsXm99:-Mt6cDG4BSWX1n5HmtEy:bwe3nsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98fa7f2f4888e9eccd8f8e2f14e616e825ea94b/pkg/server/settingswatcher/settings_watcher.go#L130)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't understand what bufferSize=0 does? Wouldn't it cause an error on the first `onValue()` callback from the rangefeed?
</blockquote></details>

No, because it's leveraging `translateEvent` and then always returning `nil` so no events get buffered unless there's storage, at which point, the buffer exists. I added more commentary.
___
*[pkg/server/settingswatcher/settings_watcher.go, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt58ejp6MztsjSqG2Al:-Mt6eiR53qdyyMFTveBY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98fa7f2f4888e9eccd8f8e2f14e616e825ea94b/pkg/server/settingswatcher/settings_watcher.go#L132)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add comment
</blockquote></details>

Done.
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go, line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt52sv00TK8fWwZhu4L:-Mt6cNiX-8l4hODoO7uD:bcp9jfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go#L142)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] should `translateEvent` be a method on `d` (in which case we would just use `d.translateEvent`)?
</blockquote></details>

heh i wrote that initially and then I felt sad about that method not being tested with the rest of the decoder methods. Done.
___
*[pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go, line 289 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt52TtG4HDackb7lEmb:-Mt6gCfeBg8gi-Ds6Rlh:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8b480bf575a41f2b27573621782904ea77b8dd/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go#L289)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] could use a short comment
</blockquote></details>

done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go, line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt56niG30dkBt4GU7ys:-Mt74Tro3eht20UoNnte:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af63f53d4db7d8ff673d143384e9ae0b5adc91d1/pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go#L83)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] not called concurrently?
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt57ZWNDeehKjXudoZl:-Mt6c_rq7hbosglW6b9A:bj3cisc) ([raw file](https://github.com/cockroachdb/cockroach/blob/af63f53d4db7d8ff673d143384e9ae0b5adc91d1/pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go#L87)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this special case could be handled in `MergeKVs`.
</blockquote></details>

Done. Took some rewording to make it sane to avoid the copy. 
___
*[pkg/server/settingswatcher/row_decoder_test.go, line 1 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt584wD87WAvVZScWif:-Mt74rBF62y0nNKMF1PX:b-pxllae) ([raw file](https://github.com/cockroachdb/cockroach/blob/02996f4a34504e0f0c1cf50eda0d6d5bcee31879/pkg/server/settingswatcher/row_decoder_test.go#L1)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[supernit] should remove this file from this commit
</blockquote></details>

Fixed the history.


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go:
s/Partial/Incremental?

irfansharif commented on pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go:
[nit] "If"

irfansharif commented on pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go:
[nit] Batch of updates? Rephrase to the singular form or refer to "events" in the plural.

irfansharif commented on pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go:
Does this need to be exported?

irfansharif replied:
Also, refers to "global store of span configs".

ajwerner replied:
I sort of like that it's exported, it makes testing other things useful, and it had clear semantics: run synchronously until there's a permanent error.

irfansharif commented on pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go:
```diff
@@ -0,0 +1,100 @@
+	case CompleteUpdate:
+		updatedData = updateKVs
+	case PartialUpdate:
+		// Note that handleUpdate is synchronously within the underlying watcher,
```
"handlUpdate is synchronously"

irfansharif reviewed (commented):
Incredibly cool. Read up to the settingswatcher change, have to come back to it and stare some more.

ajwerner commented:
@RaduBerinde is there more that you want on this other than @irfansharif's nits?

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (sorry for the delay)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-:-MuNVQ8k9X3xuAH-ycID:b-c563y0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, and @RaduBerinde)

___
*[pkg/server/settingswatcher/settings_watcher.go, line 118 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-MuDcXqn-rB4-Lg1eLos:-MuDcXqn-rB4-Lg1eLot:b7vc5bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/484b405d2a51d481acff28cfa95337096e031fa3/pkg/server/settingswatcher/settings_watcher.go#L118)):*
> ```Go
> 	}
> 	noteUpdate := func(update rangefeedcache.Update) {
> 		s.mu.Lock()
> ```

We can do the `if` outside the lock no?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-:-Mui0XkN-KI3uwpAojgz:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif and @RaduBerinde)

___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 94 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt1t0Po1fPdd5KmUWj7-r9-94:-MuZaGfI6tGThFcs-bVk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a504ac7076ec6fdd3637dc73481a908221f539/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L94)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/Partial/Incremental?
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 98 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt1t0Po1fPdd5KmUWj7-r9-98:-MuZaGxW2HZw8wS5Seib:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a504ac7076ec6fdd3637dc73481a908221f539/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L98)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] "If"
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go, line 103 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt1t0Po1fPdd5KmUWj7-r9-103:-MuZaeyzBVuo4EkUY8Sk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a504ac7076ec6fdd3637dc73481a908221f539/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L103)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Batch of updates? Rephrase to the singular form or refer to "events" in the plural.
</blockquote></details>

Done.
___
*[pkg/server/settingswatcher/settings_watcher.go, line 118 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-MuDcXqn-rB4-Lg1eLos:-MuZdji8BPDUBBqMY5di:b-1jswtt) ([raw file](https://github.com/cockroachdb/cockroach/blob/484b405d2a51d481acff28cfa95337096e031fa3/pkg/server/settingswatcher/settings_watcher.go#L118)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We can do the `if` outside the lock no?
</blockquote></details>

No, the mutex here was implicitly protecting `u`. I made that fact explicit with a little bit of refactoring. I'm glad I did it. I recall wavering on moving the updater into the struct earlier. 
___
*[pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go, line 83 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-Mt1t0PpFX3Dj_quHGLs-r12-83:-MuZf_34BHCiYiAA7vlI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4e56e9af98bb534527b228f9465b1d12dda05d/pkg/kv/kvclient/rangefeed/rangefeedcache/cache_impl_test.go#L83)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

"handlUpdate is synchronously"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/server/settingswatcher/settings_watcher.go, line 118 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-MuDcXqn-rB4-Lg1eLos:-Mukb1RE7_FfIRtJljNa:brqwhyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/484b405d2a51d481acff28cfa95337096e031fa3/pkg/server/settingswatcher/settings_watcher.go#L118)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

No, the mutex here was implicitly protecting `u`. I made that fact explicit with a little bit of refactoring. I'm glad I did it. I recall wavering on moving the updater into the struct earlier. 
</blockquote></details>

But it's not protecting update.Type.. I meant we can lock inside the if.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74612)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif)

___
*[pkg/server/settingswatcher/settings_watcher.go, line 118 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/74612#-MuDcXqn-rB4-Lg1eLos:-MukfCuVBTBSWhdwfWRZ:bgbwfyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/484b405d2a51d481acff28cfa95337096e031fa3/pkg/server/settingswatcher/settings_watcher.go#L118)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

But it's not protecting update.Type.. I meant we can lock inside the if.
</blockquote></details>

Whoops, done


<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+
